Align the turbine model and baseline power mode references with the Power Curve Schema

jonssonchristian opened the pull request:
This PR addresses issue #56.

We have previously only had a placeholder turbine model identifier in the turbine configuration data model. This is here updated to reference the turbine model and baseline power mode with reference to the Power Curve Schema.

There is an open issue on the Power Curve Schema repo to clarify the best approach to this referencing, but there has not been a response yet. I do not see the need to wait for that. We can instead refine this later if we get suggestions for improvements.

This PR does not include example power curves and associated information according to the Power Curve Schema. That can be covered in a separate issue and PR.

jonssonchristian commented on eya_def_tools/eya_def_tools/data_models/general.py:
```diff
@@ -12,7 +12,7 @@
 from eya_def_tools.data_models.base_model import EyaDefBaseModel
 
-NonEmptyStr = Annotated[str, pdt.Field(min_length=1)]
+NonEmptyStr = Annotated[str, pdt.StringConstraints(min_length=1)]
```
This is an unrelated minor refinement.

jonssonchristian commented on eya_def_tools/eya_def_tools/data_models/iea43_wra_data_model.py:
```diff
@@ -26,7 +26,10 @@
                 "Wind Task 43 WRA Data Model JSON Schema."
             ),
-            "examples": ["https://foo.com/bar/example_iea43.json"],
+            "examples": [
```
This is an unrelated minor refinement. Whilst working on the references to the Power Curve Schema, I realised it is better here to reference a real URL with example data rather than a dummy URL.

jonssonchristian commented:
Hi @momodiallosn, @bjarketol, @MeikeDittert,

I would like to test getting this PR reviewed an merged during the meeting today. I tag you here now to test if you get notified.

jonssonchristian commented:
I have expanded this PR to also cover issue #69.

momodiallosn commented on json_schema/examples/iec_61400-15-2_eya_def_example_a_XYZ-3.2_140.json:
```diff
@@ -0,0 +1,324 @@
+          ]
+        ],
+        "power_is_coefficient": false,
+        "rated_power": 3200000.0,
```
Proposal
------------
Add the turbulence intensity range (TI) for which the power curve is valid

Rationale
-----------
I've received in the past power curves  that varied with TI

jonssonchristian replied:
As noted in another comment, this follows the Power Curve Schema (IEC 61400-16), which we cannot modify within the EYA DEF. We need to raise any feedback with the Power Curve Schema (IEC 61400-16) team. That schema covers reference turbulence (as part of the design basis) but not turbulence validity ranges.

momodiallosn commented on json_schema/examples/iec_61400-15-2_eya_def_example_a_PQR169-5.8MW.json:
```diff
@@ -0,0 +1,333 @@
+            "wind_speed": 24.0
+          },
+          {
+            "kind": "high_cut_out",
```
Seems to be a duplicate of line 86

jonssonchristian replied:
No, both are `high_cut_out`, but they have different averaging period and different wind speed values. In this example there are separate cut-outs on 10-minute average data, 30-second average data and 3-second average data.

momodiallosn commented on eya_def_tools/eya_def_tools/utils/doc_utils/eya_def_erdantic.py:
```diff
@@ -181,8 +181,8 @@ def _get_abbreviated_type_name(
     match field_name:
         case "measurement_stations" | "reference_meteorological_datasets":
             return "Optional[list[IEATask43WraDataModel]]"
-        case "turbine_models":
```
why did you remove case "turbine_models": in simply not replacing it by case "turbine".
It seems that is available in https://github.com/IEC-61400-15/eya_def/blob/552043c7dc8d45dedb42a96b83bc2611558ea941/json_schema/examples/iec_61400-15-2_eya_def_example_a_ABC165-5.5MW.json
line 292

jonssonchristian replied:
The key `turbine_models` was renamed to `power_curves` at the top level of the schema, so renamed here as well. Since the schema is called the Power Curve Schema, I thought `power_curves` more consistent here than `turbine_models`.

What you reference is within the Power Curve Schema, and we do not include that breakdown in the diagrams. The code here is just for the labelling of types for our diagrams.

momodiallosn commented on eya_def_tools/eya_def_tools/tests/test_schema_validation.py:
```diff
@@ -82,6 +83,46 @@ def test_validate_iea43_wra_data_model(
+    )
+
+
+def test_json_schema_validate_power_curve_document_a(
```
Refactorisation proposal:
Instead of having test_json_schema_validate_power_curve_document_a, test_json_schema_validate_power_curve_document_b and test_json_schema_validate_power_curve_document_c

Maybe using one single function that can take a dict or a list of dict (proposal to be tested)

def test_json_schema_validate_power_curve_document(
    power_curve_document_or_documents: Union[dict[str, Any], List[dict[str, Any]]],
    iec61400_16_power_curve_schema_json_schema: dict[str, Any],
) -> List[]:

    if isinstance(power_curve_document_or_documents, dict):
        # Handle a single document
            results=jsonschema.validate(instance=power_curve_document_or_documents, 
                                schema=iec61400_16_power_curve_schema_json_schema)
    else:
        # Handle a list of documents
        results = []
        for doc in power_curve_document_or_documents:
                results.append(jsonschema.validate(instance=doc, schema=iec61400_16_power_curve_schema_json_schema))

        return results

jonssonchristian replied:
You are correct I can parameterise over the cases. There are just some limitations in the `pytest` library that you cannot parameterise over different fixtures. It is better to have one test for each case, so you get clearer result data on what failed. I will update this code.

jonssonchristian replied:
I have now updated this to parameterise the test.

momodiallosn reviewed (commented):
Minor improvements:
- including TI range of power curves
-refacto of code for test

jonssonchristian commented:
> Minor improvements:
> 
> * including TI range of power curves
>   -refacto of code for test

The power curve specifications are covered within the Power Curve Schema. The turbulence conditions for which the turbine is designed is covered under design bases, see here:

https://github.com/octue/power-curve-schema/blob/2b95e3b853d5f994ad06686e530a156caf298471/power-curve-schema/schema.json#L724

There are no upper and lower turbulence limits in that schema. If we think there should be, we should suggest it to the Power Curve Schema (IEC 61400-16) team.

jonssonchristian commented on json_schema/iec_61400-15-2_eya_def.schema.json:
```diff
@@ -3006,7 +3013,8 @@
                 "hub_height",
-                "turbine_model_id"
+                "turbine_model",
+                "baseline_operating_mode"
```
Both the turbine and wind farm objects already have the optional attribute `restrictions` (see below), with a basic data model). It is optional since not all projects have restrictions. We discussed this at length at an earlier stage and concluded that a full quantitative description of operational modes is beyond the scope of the first version of the EYA DEF, but may be added later.

![image](https://github.com/user-attachments/assets/193636e2-88e5-4395-8795-8a83ca48d251)

jonssonchristian commented:
Hi @momodiallosn, thank you very much for your input and feedback. I have now addressed all of your comments and pushed one update to the code. Let me know if you have any further comments or questions. We can also discuss during the upcoming meeting.
